### DISCUSSION TITLE
Handful of Ideas

### DISCUSSION THREAD

**Discussion by jbwill36** [TechnitiumSoftware/DnsServer#843] [Ideas]
I've been using Technitium off and on for several years and currently use it for a Telecom stack operating in all 50 US states, so it is a solution that I believe in, so thank you much for open-sourcing and maintaining Technitium. The following are some features that would be awesome to see:
1. Path& Port based forwarding for places where there is a need to route multiple different domains to port values, or paths like Caddy but also with ports. I have multiple sbcs that serve different domains based on location but have IP constraints in that region for whatever reason. Being able to use a reverse proxy for domain names as well as ports would be amazing. That way I could point example.com to 192.168.1.1 and sub.example.com to 192.168.2.1 or use port values like example.com:8888 to 192.168.1.1 and new.example.com:8889 using the same public address for both. especially if you wired it with the ability to resolve or forward HTTP-01 requests. 
2. External Database Support. I have a lot of CDR data to parse and have database farms. Some of my clients also host other services with us, being able to maintain compliance and access logs for their needs would be awesome. 
3. Import an Entire config in JSON. This would just make life easier all around having a single document to manage. I could send some examples if you wish to discuss this in terms of feasibility because I have some ideas on it. 
4. Plugin documentation for 3rd party plugins. People would turn out some pretty cool stuff if this existed. I would put some resources toward that and I would also be willing to create some documentation and maintain it going forward with a little bit of education on some of the existing plugins that way you don't have to, 
Thanks again for all you do and for this awesome little DNS server app. You guys are awesome.

**Comment by jbwill36** [comment]
it would also be awesome to have the ability to load a global settings file. That would set the defaults for new zones created, including nameservers, specific DNS records, TSIG assignments w/Record types, as well as importing a global zone file with all these values in it. That would make deployment fast, easy, consistent, especially if it could pull from a private git. Turn on the app, provide a key, url to git, go grab some coffee and enjoy the rest of your day.

**Comment by ShreyasZare** [comment]
Thanks for the suggestions and compliments. Good to know that the DNS server is working well for your requirements.

> Path& Port based forwarding for places where there is a need to route multiple different domains to port values, or paths like Caddy but also with ports. I have multiple sbcs that serve different domains based on location but have IP constraints in that region for whatever reason. Being able to use a reverse proxy for domain names as well as ports would be amazing. That way I could point example.com to 192.168.1.1 and sub.example.com to 192.168.2.1 or use port values like example.com:8888 to 192.168.1.1 and new.example.com:8889 using the same public address for both. especially if you wired it with the ability to resolve or forward HTTP-01 requests.

DNS server cannot work with path or port since that is done at the HTTP protocol level. The DNS server can only answer for domain names and knows nothing about the protocol or the HTTP URL being used.

> External Database Support. I have a lot of CDR data to parse and have database farms. Some of my clients also host other services with us, being able to maintain compliance and access logs for their needs would be awesome.

If you are asking about query logging on external database then there is plan to add DNS app to do that. Currently there is only one DNS app that works with a local sqlite db.

> Import an Entire config in JSON. This would just make life easier all around having a single document to manage. I could send some examples if you wish to discuss this in terms of feasibility because I have some ideas on it.

The software architecture is a bit different so having json config file wont be feasible. Its recommended to use HTTP API to manage the server's settings which can be used with small bash scripts using curl.

> Plugin documentation for 3rd party plugins. People would turn out some pretty cool stuff if this existed. I would put some resources toward that and I would also be willing to create some documentation and maintain it going forward with a little bit of education on some of the existing plugins that way you don't have to,

There is lack of documentation for the project due to time constraints and may take some time for that to be made available. For writing custom apps/plugins, you can refer to the existing code on GitHub to know how its implemented. If any help is required, you can always write an email to support.

> it would also be awesome to have the ability to load a global settings file. That would set the defaults for new zones created, including nameservers, specific DNS records, TSIG assignments w/Record types, as well as importing a global zone file with all these values in it. That would make deployment fast, easy, consistent, especially if it could pull from a private git. Turn on the app, provide a key, url to git, go grab some coffee and enjoy the rest of your day.

As said earlier, the architecture is a bit different for using config files. Its recommended to use HTTP API to automate any task using scripts. For zones, there is already support to import/export records in standard zone file format which can be used as a template to add records in bulk for new zones.

Let me know if you have any more queries or feedback.
